url8417b1b213_Ratchette_Rummikub_tgz-pJ8-rummikub_RummikubJ8/src/rummikub/Move.java

```diff
@@ -0,0 +1,163 @@
+package rummikub;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+
+// FIXME - this was thrown in at the last second
+public class Move {
+	public Hand handBefore;
+	public Hand handAfter;
+	public Hand usedTiles;
+	
+	public ArrayList<Meld> boardBefore;
+	public ArrayList<Meld> boardAfter;
+	
+	public HashMap<Tile, ArrayList<Position>> positionsBefore;
+	public HashMap<Tile, ArrayList<Position>> positionsAfter;
+	public HashMap<Position, Position> movement;
+	
+	
+	public Move(Hand startHand, Hand endHand, ArrayList<Meld> startBoard, ArrayList<Meld> endBoard) throws Exception{
+		handBefore = startHand;
+		handBefore.sortByColour();
+		
+		handAfter = endHand;
+		handAfter.sortByColour();
+		
+		boardBefore = startBoard;
+		Collections.sort(boardBefore);
+		boardAfter = endBoard;
+		Collections.sort(boardAfter);
+		
+		calculateUsedTiles();
+		removeUntouchedMelds();
+		
+		System.out.println("\n--------------------------------------------------------\n");
+		System.out.println("Starting Tiles");
+		displayMelds(boardBefore);
+		
+		System.out.println("\nEnding Tiles");
+		displayMelds(boardAfter);
+		
+		calculatePositions();
+		displayMoves();
+	}
+
+	private void calculateUsedTiles() throws Exception{
+		ArrayList<Tile> tilesBefore;
+		ArrayList<Tile> tilesAfter;
+		
+		tilesBefore = handBefore.getTiles();
+		tilesAfter = handAfter.getTiles();
+		
+		for (Tile tile : tilesAfter)
+			tilesBefore.remove(tile);
+		
+		usedTiles = new Hand(new ArrayList<Tile>(tilesBefore));
+	}
+	
+	private void removeUntouchedMelds(){
+		ArrayList<Meld> startingMelds = new ArrayList<Meld>(boardBefore);
+		
+		for(Meld meld : startingMelds){
+			if(boardAfter.contains(meld)){
+				boardBefore.remove(meld);
+				boardAfter.remove(meld);
+			}
+		}
+	}
+	
+	private void calculatePositions() throws Exception{
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 78-89, lines 93-104, lines 105-123 into separate methods.